finality-aleph/src/aggregation/aggregator.rs

```diff
@@ -14,29 +14,112 @@ use std::{
     marker::PhantomData,
 };
 
-pub(crate) enum AggregatorError {
+pub enum AggregatorError {
+    LastHashPlaced,
+    NoHashFound,
+    DuplicateHash,
+}
+
+pub enum IOError {
     NetworkChannelClosed,
+    AggregatorError(AggregatorError),
 }
 
+pub type AggregatorResult = Result<(), AggregatorError>;
+pub type IOResult = Result<(), IOError>;
+
 /// A wrapper around an `rmc::Multicast` returning the signed hashes in the order of the [`Multicast::start_multicast`] calls.
 pub struct BlockSignatureAggregator<
+    'a,
+    H: Hash + Copy,
+    MK: MultiKeychain,
+    MS: Multisigned<'a, SignableHash<H>, MK>,
+> {
+    signatures: HashMap<H, MK::PartialMultisignature>,
+    hash_queue: VecDeque<H>,
+    last_hash_placed: bool,
+    started_hashes: HashSet<H>,
+    metrics: Option<Metrics<H>>,
+    marker: PhantomData<&'a MS>,
+}
+
+impl<'a, H: Copy + Hash, MK: MultiKeychain, MS: Multisigned<'a, SignableHash<H>, MK>>
+    BlockSignatureAggregator<'a, H, MK, MS>
+{
+    pub(crate) fn new(metrics: Option<Metrics<H>>) -> Self {
+        BlockSignatureAggregator {
+            signatures: HashMap::new(),
+            hash_queue: VecDeque::new(),
+            last_hash_placed: false,
+            started_hashes: HashSet::new(),
+            metrics,
+            marker: PhantomData,
+        }
+    }
+
+    fn on_start(&mut self, hash: H) -> AggregatorResult {
+        if !self.started_hashes.insert(hash) {
+            return Err(AggregatorError::DuplicateHash);
+        }
+        if let Some(metrics) = &self.metrics {
+            metrics.report_block(hash, std::time::Instant::now(), Checkpoint::Aggregating);
+        }
+        self.hash_queue.push_back(hash);
+
+        Ok(())
+    }
+
+    pub(crate) fn notify_last_hash(&mut self) {
+        self.last_hash_placed = true;
+    }
+
+    fn on_multisigned_hash(&mut self, multisigned_hash: MS) -> AggregatorResult {
```

Comment:
This shouldn't be returning a result if it cannot ever return an error – I thought this was a mistake in a logging message about network failing below, but the problem actually is here.